tools/mpremote: Add manifest function.

andrewleech opened the pull request:
Add manifest command to mpremote which will compile a manifest.py from the specified folder, eg:

``` bash
      $ MPY_DIR=../../micropython 
      $ PORT_DIR=../../micropython/ports/esp32
      $ mpremote manifest .
  ```
  This will assemble / mpy-cross everything specified in the manifest.py into the folder `_manifest`.
  If the current folder is also mounted, this folder will automatically be added to the path, eg:

``` bash
      $ mpremote manifest . mount .
```
  
Alternatively, the built folder can by copied / synched to the device, in which case 
  the copied folder will be added to the path:

``` bash
      $ mpremote manifest -s . 
```
  A soft-reset will re-process the manifest file to include any local updates allowing fast development of local files in a larger project without needing to copy / import-over-mount all files every reset.


Ideally, if the `mpy_cross` python package is installed from pypi, the local `MPY_DIR=../../micropython` variables will not need to be specified (unless your local manifest uses them directly`

andrewleech commented:
mpremote is also being packaged to wheel in github-actions in this PR, can be tested with
```
wget https://github.com/andrewleech/micropython/suites/5119362628/artifacts/153770325
unzip mpremote.zip
pip install mpremote*.whl
```

andrewleech commented on tools/mpremote/mpremote/console.py:
```diff
@@ -11,8 +11,13 @@
+        if hasattr(self.infile, "raw"):
+            self.infile = self.infile.raw
+        if hasattr(self.outfile, "raw"):
+            self.outfile = self.outfile.raw
```
When running mpremote in the vscode terminal on OSX the sys.stdout.buffer did not have the raw attribute - it appeared to work fine without it.

andrewleech commented:
I still want to do some more work here to speed up the sync process, make it as efficient as I can. 

With a remote littlefs partition on esp32 even just doing recursive file listing can be painfully slow ( ~ 12 seconds for 128 files across 28 folders ), not sure how much could be done to improve this.

One of the main things I'd like to fix is more related to `fs cp -r` than the manifest. 
When copying a folder from computer over an existing remote folder, existing files will be overwritten however remote files/folders contained within that have since been deleted on computer will be left on the device. This way you can have stale content causing unexpected results.
On linux, a `cp -r` will entirely replace an existing folder unless you add the `-u` flag.

dpgeorge commented:
I want to look into general file transfer speed using `mpremote`.  It should be possible to improve the speed for files going in to the device.



> On linux, a `cp -r` will entirely replace an existing folder unless you add the `-u` flag.

+1

andrewleech commented:
I haven't looked at the actual file transfer speed over the serial port, however I've got re-copying folders working much faster with either a sha or (size + mtime) check on remote vs local filesystems.

The best speed so far is by scanning the remote folder tree into a dict and passing that back to mpremote to check against local folder; this can be used to delete extra remote files and not transfer ones that aren't needed. 

Most of the time for this operation on a mostly unchanged folder is just in a recursive walk in micropython, even with just one os.stat per file it can be quite slow. Might not be able to do much about that though, no way to know if there are extra files to delete without doing a os.listdir and then os.stat on each one to know if it's a dir or a file...

dpgeorge commented:
> no way to know if there are extra files to delete without doing a os.listdir and then os.stat on each one to know if it's a dir or a file...

try `uos.ilistdir()`

andrewleech commented:
Yes of course, I did forget about the difference between the two. I'll try that later and see how much improvement it makes

peterhinch commented:
Is there any progress on https://github.com/micropython/micropython/issues/7731?

andrewleech commented:
> Is there any progress on #7731?

I haven't run into this issue myself so hadn't seen that ticket. At least it looks reasonably easy to reproduce, though there could be a few issues in that mix too

andrewleech commented:
For now I've trimmed this PR back down to just the manifest command, no sync. This command combined with mount is enough on many devices (ones with native usb at least) to allow fast development of projects which use a manifest file to organise the files / libraries. 

I'll continue to work on `sync` / `deploy` command/s, along with any file transfer speed improvements & fixes in a follow up PR.

Josverl commented on tools/mpremote/pyproject.toml:
```diff
@@ -1,6 +1,7 @@
     "setuptools>=42",
-    "wheel"
+    "wheel",
+    "mpy-cross",
```
should 'mpy-cross' be pinned to a specific version, or is it intentionally un-versioned ?

andrewleech replied:
My intention at the time was to be un-pinned, to always use latest (because I generally always use latest micropython).
However now, there's efforts under way to have the mpy-cross package built / managed as part of this repo and integrate it into mpremote in a cleaner and more compatible way.

pi-anl committed:
tools/mpremote: Add manifest command.

Add a `manifest` command that processes a MicroPython manifest.py file,
compiles .py files to .mpy using mpy-cross, and places the output in a
local build directory.  When combined with `mount`, the compiled modules
are added to the device's sys.path and are automatically re-built on
soft-reset.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>
